crates/dbsp/src/trace/layers/file/column_layer/builders.rs

```diff
@@ -0,0 +1,164 @@
+    ) where
+        F: Fn(&<<Self::Trie as Trie>::Cursor<'a> as Cursor<'a>>::Key) -> bool,
+    {
+        let mut cursor = other.cursor_from(lower, upper);
+        while cursor.valid() {
+            let item = cursor.current_item();
+            if filter(&item.0) {
+                self.0.write0((&item.0, &item.1)).unwrap();
+            }
+            cursor.step();
+        }
+    }
+
+    fn push_merge_retain_keys<'a, F>(
+        &'a mut self,
+        mut cursor1: <Self::Trie as Trie>::Cursor<'a>,
+        mut cursor2: <Self::Trie as Trie>::Cursor<'a>,
+        filter: &F,
+    ) where
+        F: Fn(&<<Self::Trie as Trie>::Cursor<'a> as Cursor<'a>>::Key) -> bool,
+    {
+        while cursor1.valid() && cursor2.valid() {
+            match cursor1.current_key().cmp(cursor2.current_key()) {
+                Ordering::Less => {
+                    let item = cursor1.current_item();
+                    if filter(&item.0) {
+                        self.0.write0((&item.0, &item.1)).unwrap();
+                    }
+                    cursor1.step();
+                }
+                Ordering::Equal => {
+                    let mut sum = cursor1.current_diff().clone();
+                    sum.add_assign_by_ref(cursor2.current_diff());
+
+                    if !sum.is_zero() && filter(cursor1.current_key()) {
+                        self.0.write0(((cursor1.current_key()), &sum)).unwrap();
+                    }
+                    cursor1.step();
+                    cursor2.step();
+                }
+
+                Ordering::Greater => {
+                    let item = cursor2.current_item();
+                    if filter(&item.0) {
+                        self.0.write0((&item.0, &item.1)).unwrap();
+                    }
+                    cursor2.step();
+                }
+            }
+        }
+
+        while let Some(item) = cursor1.take_current_item() {
+            if filter(&item.0) {
+                self.0.write0((&item.0, &item.1)).unwrap();
+            }
+        }
+        while let Some(item) = cursor2.take_current_item() {
+            if filter(&item.0) {
+                self.0.write0((&item.0, &item.1)).unwrap();
+            }
+        }
+    }
+}
+
+impl<K, R> TupleBuilder for FileColumnLayerBuilder<K, R>
+where
+    K: DBData,
+    R: DBWeight,
+{
+    type Item = (K, R);
+
+    fn new() -> Self {
+        Self(Writer1::new(tempfile().unwrap(), Parameters::default()).unwrap())
+    }
+
+    fn with_capacity(_capacity: usize) -> Self {
+        Self::new()
+    }
+
+    fn reserve_tuples(&mut self, _additional: usize) {}
+
+    fn tuples(&self) -> usize {
+        self.0.n_rows() as usize
+    }
+
+    fn push_tuple(&mut self, item: (K, R)) {
+        self.0.write0((&item.0, &item.1)).unwrap();
+    }
+}
+
+impl<K, R> SizeOf for FileColumnLayerBuilder<K, R>
+where
+    K: DBData,
+    R: DBWeight,
+{
+    fn size_of_children(&self, _context: &mut size_of::Context) {
+        // XXX
```

Comment:
This whole size-of API might needs some reconsideration anyways with persistence @ryzhyk